integr8ly/integreatly-operator
# INTLY-8143 Modify RHMIConfig upgrade scheduling logic

# Description

> Link to JIRA: https://issues.redhat.com/browse/INTLY-8143

Modify upgrade scheduling logic to be based uniquelly on two new spec values,
`notBeforeDays` and `waitForMaintenance`. The upgrade will be scheduled at
n + m days after the upgrade is made available, being:

* n the value of `notBeforeDays`
* m the remaining days until the next maintenance window after n days if
  `waitForMaintenance` is set to true, or zero otherwise.

The default value of these fields is 7 for `notBeforeDays`, and `true` for
`waitForMaintenance`

Update validation and defaulting logic.
Update unit tests.

## Verification steps

In order to fully verify this, you must test the upgrade from the current version (with the old CRD fields) to the new one, to verify that existent RHMIConfig CRs are well handled during and after the upgrade:

1. Install the previous operator version through OLM (2.3.0)
2. Build an image from this PR (2.3.1) and set up an upgrade to this operator version
3. Edit the RHMIConfig to approve the upgrade: set the `alwaysImmediately` field to `true`
4. Wait for the upgrade to be automatically approved and the new version to be rolled out
5. Verify that after a few minutes, the RHMIConfig is automatically modified to the new version, with the default values of the upgrade spec (`{ notBeforeDays: 7, waitForMaintenance: true }`)
6. Verify the validating and mutating webhooks:
    1. Attempt to set an invalid value for notBeforeDays (non integer, or less than 0)
    2. Attempt to set an invalid value for `waitForMaintenance` (not `true` or `false`)
    3. Verify that the update is denied
    4. Set valid values for the fields and verify that the update is allowed
    5. Attempt to remove the fields or leave them empty and verify that the old value is set back
7. Set up an upgrade to a newer version (2.3.2)
8. Verify that the upgrade schedule status is set based on the spec
9. Change the spec to `{ notBeforeDays: 0, waitForMaintenance: false }`
10. Verify that the upgrade is approved

## Type of change




## Checklist
